### ISSUE TITLE
Feature (request): throw more than on error at time

### ISSUE THREAD

**Issue by dpturibio** [textX/textX#202]
While developing a Language recognizer for a specific language that my team created for Magnetic resonance, that will work in Spyder Ide, I found that when I compile my model from code, and something does not match, only one error is thrown.

```
    try:
        meta.model_from_str(code)
        print("model ok")
    except TextXError as e:
        print("Error in model: ", e)
```

I would like to propose that when exist more than one error in the code, that a vector with the subsequent errors can be thrown, as this characteristic is essential for a language recognizer in my point of view.

**Comment by goto40** [comment]
Here, I try to **summarize what problems can occur**, while loading a model:

 * **Parser** problems (syntax) 
 * **Reference resolution** problems ("object not found")
 * **Object processor / validation** problems (custom semantic checks)

Any problem can either **(1)** allow to collect further problems (new behavior possibly included in this *new feature*) or to **(2)** stop the entire process of model loading (aka throw the error, *current behavior*, possibly together with previously collected errors).

**Task A**. It has to be discussed which types of problems support being collected (and thrown/reported together), and which not ('severe errors' which imply to stop the entire process of model loading).

**Task B.** Then, we need to discuss how (the new method of) reporting and collecting errors could be included in the software design of textX.

**Comment by igordejanovic** [comment]
For reporting multiple syntax error in a single file a parser should be able to recover from an error and continue. To do error recovery parser state must be manipulated.
Generally it is very hard to do error recovery when the most of the parser state is kept implicitly on the call stack and that is the case for practically all PEG parser (at least to my knowledge) as the most straightforward way to implement parsers for PEG grammar is by recursive descent (e.g. packrat parsing in Arpeggio). There is ongoing research in the area of error recovery for PEG parsers so we might see if the current findings could be used.

**Comment by goto40** [comment]
Errors while resolving references could reported together.

Errors in object processors remain a point to clarify. It could help to distinguish between 'validators' and modifying processors. Errors from validators could be collected. Errors from modifying processors should probably stop the loading process... However, we do not distinguish between processor types.

**Comment by igordejanovic** [comment]
I agree. Error in modifying processors might impact further processing which might yield multiple false errors. Syntax error recovery is usually a heuristic algorithm so is very similar in that regard. Usual way to deal with the problem is to limit the number of errors reported. Say, we report up to 10 errors and stop loading if there are more.

**Comment by Geniedesalpages** [comment]
I would also like to be able to "skip" a syntax error, by adding some kind of resolution function `f` that is called when for instance a `';'` is expected. `f` then feeds the parser a virtual ';' and the parser tries to resume. Given the the current parsing stack and character more complex behaviour would be possible (`,` replace with `;`).

**Comment by igordejanovic** [comment]
Would be nice to have but it is not easy in PEG packrat parser. For that to happen parser would have to restore its state where it expects `;` in the input stream and restoring state with recursive descent parsers if possible would be very hacky I guess. I don't think that is something that will be implemented soon in textX due to the amount of work necessary for both research and implementation. That change would have to be done in Arpeggio so if anyone has a time and will to try to solve it I would be very happy to help.

If error recovery is a must for you use case I suggest you to take look at some other parsers that provide that feature. For example, I also work on [parglare parser](https://github.com/igordejanovic/parglare/) which is a bottom-up (G)LR parser which inherits some ideas from textX and which has support for error recovery and is way more powerful than textX in terms of features but it is still not that straightforward and easy to use, although some things are much easier, like implementing expression languages.
In order to use it you should be familiar with LR shift-reduce parsing.

**Comment by Geniedesalpages** [comment]
Ok, that makes sense, thanks!
